### PR TITLE
chore: use environment vars to configure global proxy settings

### PR DESCRIPTION
**Pull request by hessjcg**
Use environment variables instead of CLI flags to configure the global proxy settings
for proxy containers.

### CODE REVIEW

File: internal/workload/podspec_updates.go [hunk 1]
```diff
@@ -544,25 +556,31 @@ func (s *updateState) applyContainerSpec(p *cloudsqlapi.AuthProxyWorkload, c *co
...
 
 // updateContainerEnv applies global container state to all containers
 func (s *updateState) updateContainerEnv(c *corev1.Container) {
 	for i := 0; i < len(s.mods.EnvVars); i++ {
 		var found bool
-		operatorEnv := s.mods.EnvVars[i].OperatorManagedValue
+		v := s.mods.EnvVars[i]
+		operatorEnv := v.OperatorManagedValue
+
+		// This envvar is not global and doesn't apply to this container, skip it.
```
Comment (enocom) [inline internal/workload/podspec_updates.go:580]: What does global mean when applied to an envvar? I though all envvars were global.

File: internal/workload/podspec_updates.go [hunk 2]
```diff
@@ -605,12 +620,12 @@ func (s *updateState) addHealthCheck(_ *cloudsqlapi.AuthProxyWorkload, c *corev1
...
-	cliArgs = append(cliArgs,
-		fmt.Sprintf("--http-port=%d", port),
-		"--http-address=0.0.0.0",
-		"--health-check",
-		"--structured-logs")
-	return cliArgs
+	s.addProxyContainerEnvVar(p, "CSQL_PROXY_HTTP_PORT", fmt.Sprintf("%d", port))
+	s.addProxyContainerEnvVar(p, "CSQL_PROXY_HTTP_ADDRESS", "0.0.0.0")
+	s.addProxyContainerEnvVar(p, "CSQL_PROXY_HEALTH_CHECK", "true")
+	s.addProxyContainerEnvVar(p, "CSQL_PROXY_STRUCTURED_LOGS", "true")
```
Comment (enocom) [inline internal/workload/podspec_updates.go:626]: Why configure logs here? From the method name (addHealthCheck), I would expect only health check config to be here.

File: internal/workload/podspec_updates.go [hunk 3]
```diff
@@ -330,36 +334,44 @@ func (s *updateState) addPort(p int32, containerName string, n types.NamespacedN
...
-		if s.mods.EnvVars[i].OperatorManagedValue.Name == ev.Name {
-			old := s.mods.EnvVars[i].OperatorManagedValue
-			s.mods.EnvVars[i].OperatorManagedValue = ev
-			if old.Value != ev.Value {
-				s.addError(cloudsqlapi.ErrorCodeEnvConflict,
-					fmt.Sprintf("environment variable named %s already exists", ev.Name), p)
-			}
+		oldEnv := s.mods.EnvVars[i]
+		// if the values don't match and either one is global, or its set twice
+		if oldEnv.OperatorManagedValue.Name == v.OperatorManagedValue.Name &&
```
[the remark this answers is not in the document]
Comment (hessjcg) [reply internal/workload/podspec_updates.go]: Done.

File: internal/workload/podspec_updates.go [hunk 4]
```diff
@@ -330,36 +334,44 @@ func (s *updateState) addPort(p int32, containerName string, n types.NamespacedN
...
+			oldEnv.OperatorManagedValue.Value != v.OperatorManagedValue.Value &&
+			(oldEnv.ContainerName == "" || v.ContainerName == "" || oldEnv.ContainerName == v.ContainerName) {
+			s.addError(cloudsqlapi.ErrorCodeEnvConflict,
+				fmt.Sprintf("environment variable named %s already exists, old: %v new: %v, old value: %v new value: %v",
```
[the remark this answers is not in the document]
Comment (hessjcg) [reply internal/workload/podspec_updates.go]: Good catch. I've reduced it to just the EnvVar name.

File: internal/workload/podspec_updates.go (hunk 1 above)
Comment (hessjcg) [reply internal/workload/podspec_updates.go]: I rewrote the comment for clarity, I hope.

File: internal/workload/podspec_updates.go (hunk 2 above)
Comment (hessjcg) [reply internal/workload/podspec_updates.go]: I moved it out of addHealthCheck()  and up into the updateContainer() method.

File: internal/workload/podspec_updates_test.go [hunk 5]
```diff
@@ -714,6 +697,21 @@ func TestProxyCLIArgs(t *testing.T) {
 			// test that port cli args are set correctly
 			assertContainerArgsContains(t, csqlContainer.Args, tc.wantProxyArgContains)
 
+			// Test that workload has the right env vars
+			for wantKey, wantValue := range tc.wantWorkloadEnv {
+				containerName := "busybox"
+				if strings.HasPrefix(wantKey, "CSQL_PROXY_") {
```
[the remark this answers is not in the document]
Comment (hessjcg) [reply internal/workload/podspec_updates_test.go]: Good catch. Fixed.

File: internal/workload/podspec_updates_test.go [hunk 6]
```diff
@@ -714,6 +697,21 @@ func TestProxyCLIArgs(t *testing.T) {
...
+					containerName = csqlContainer.Name
+				}
+				gotEnvVar, err := findEnvVar(wl, containerName, wantKey)
+				if err != nil {
+					t.Error(err)
+					logPodSpec(t, wl)
```
[the remark this answers is not in the document]
Comment (hessjcg) [reply internal/workload/podspec_updates_test.go]: Fixed.

**Review by hessjcg — COMMENTED** [review]
Ready for another review.

File: internal/workload/podspec_updates.go [hunk 7]
```diff
@@ -254,8 +253,13 @@ type workloadMods struct {
...
 
-func (s *updateState) addInUsePort(p int32, containerName string) {
-	s.addPort(p, containerName, types.NamespacedName{}, "")
+func (s *updateState) addInUsePort(p int32, proxy *cloudsqlapi.AuthProxyWorkload) {
+	var name types.NamespacedName
+	if proxy != nil {
+		name.Namespace = proxy.Namespace
+		name.Name = proxy.Name
+	}
+	s.addPort(p, name, "")
```
Comment (enocom) [inline internal/workload/podspec_updates.go:262]: Why empty string here? Can we add a comment to clarify?

File: internal/workload/podspec_updates.go [hunk 8]
```diff
@@ -330,36 +334,57 @@ func (s *updateState) addPort(p int32, containerName string, n types.NamespacedN
...
 		s.mods.Ports = append(s.mods.Ports, mp)
 	}
-	if containerName != "" && !strings.HasPrefix(containerName, ContainerPrefix) {
-		mp.OriginalValues[containerName] = p
-	}
+}
 
+func (s *updateState) addProxyContainerEnvVar(p *cloudsqlapi.AuthProxyWorkload, k, v string) {
+	s.addEnvVar(p, managedEnvVar{
+		Instance:             dbInst(p.Namespace, p.Name, ""),
```
Comment (enocom) [inline internal/workload/podspec_updates.go:346]: Again the empty string confuses me here.

File: internal/workload/podspec_updates.go [hunk 9]
```diff
@@ -330,36 +334,57 @@ func (s *updateState) addPort(p int32, containerName string, n types.NamespacedN
...
-		if s.mods.EnvVars[i].OperatorManagedValue.Name == ev.Name {
-			old := s.mods.EnvVars[i].OperatorManagedValue
-			s.mods.EnvVars[i].OperatorManagedValue = ev
-			if old.Value != ev.Value {
-				s.addError(cloudsqlapi.ErrorCodeEnvConflict,
-					fmt.Sprintf("environment variable named %s already exists", ev.Name), p)
-			}
+		oldEnv := s.mods.EnvVars[i]
+		// if the values don't match and either one is global, or its set twice
+		if isEnvVarConflict(oldEnv, v) {
```
Comment (enocom) [inline internal/workload/podspec_updates.go:363]: Nice.

**Review by enocom — APPROVED** [review]

File: internal/workload/podspec_updates.go (hunk 7 above)
Comment (hessjcg) [reply internal/workload/podspec_updates.go]: I updated the function names here for clarity.

File: internal/workload/podspec_updates.go (hunk 8 above)
Comment (hessjcg) [reply internal/workload/podspec_updates.go]: Same here.

**Pull request merged**